# Template for Solving Models

Every basic Gurobi optimization model can be constructed with the code template shown below.  This template is provided as a starting point for each problem formulations below. Feel free to use it for your work.

```python
import gurobipy as gp
from gurobipy import GRB

''' Import or define problem data '''

''' Create Gurobi model object '''
m = gp.Model('') # insert model name in quotes

''' Specify whether model is maximized or minimized   (model sense) '''
m.ModelSense = GRB.MAXIMIZE

''' Specify optimization parameter settings, if desired '''
# m.setParam('TimeLimit', 7200)

''' Create decision variables and update model'''
# Use m.addVar(), m.addVars() or m.addMVar() here
m.update()

''' Create objective function and update model '''
m.setObjective()
m.update()

''' Create constraints and update model '''
# Use m.addConstr(), m.addLConstr(), m.addConstrs(), or m.addMConstr() here
m.update()

''' Optimize model '''
m.optimize()

''' Print results '''

```

# Exploring the Galaxy Toys Example

We are now going to explore the Galaxy Toys example a bit more. Let's first see how to create and solve the model using Gurobi. Then, how can we get the sensitivity analysis information out?

In [107]:
# import the two Gurobi parts we need
import gurobipy as gp
from gurobipy import GRB
# import pandas for help in functions
import pandas as pd

In [108]:
# Define a function that will return sensitivity analysis for the variables
def sa_vars(the_vars):
    """
    This function takes the variables from a solved model
    and returns a pandas DataFrame where the index is the
    variable name and the columns are the (1) the resulting
    value of the variable, (2) the reduced cost, (3) the
    objective function coefficient, (4) the low end of the
    range of optimality, and (5) the high end of the range
    of optimality

    Parameters
    ===============
    the_vars: the variables of the solved model (e.g., the
              result of m.getVars())

    Returns
    ===============
    A pandas DataFrame object
    """
    sa = {}
    for v in the_vars:
        sa[v.VarName] = [v.X, v.RC, v.Obj, v.SAObjLow, v.SAObjUp]

    return pd.DataFrame.from_dict(sa,
                       orient='index',
                       columns=['final_value', 'reduced_cost', 'obj_coef', 'range_opt_low', 'range_opt_high'])

In [109]:

def sa_constrs(the_constrs):
    """
    This function takes the constraints from a solved model
    and returns a pandas DataFrame where the index is the
    constraint name and the columns are the (1) indication 
    if the constraint is binding or non-binding, (2) the
    value of the constraint, (3) the RHS of the constraint,
    (4) the slack, (5) the shadow price, (6) the low end of
    the range of feasibility, and (5) the high end of the
    range of feasibility

    Parameters
    ===============
    the_constrs: the constraints of the solved model (e.g., the
              result of m.getConstrs())

    Returns
    ===============
    A pandas DataFrame object
    """
    sa = {}
    for c in the_constrs:
        binding = 'binding'
        if c.Slack > 0.00001:
            binding = 'non-binding'
        sa[c.constrName] = [binding, c.RHS-c.Slack, c.RHS, c.Slack, c.pi, c.SARHSLow, c.SARHSUp]

    return pd.DataFrame.from_dict(sa,
                        orient='index',
                        columns=['binding?', 'final_value', 'RHS', 'slack', 'shadow_price', 'range_feas_low', 'range_feas_high'])

## Very Simple Approach

We are going to start with the most basic and "simple" way to create and solve our small LP using Gurobi. We will do it steps, with each step in its own code cell. (You can, of course, put all these steps into a single code cell.) NOTE: We should probably put a `try` block around all of this code to catch any runtime errors. 

## Formulation 

Recall that our formulation looked like this:

| | | |
| --- | --- | --- |
| Let | | |
| $x_{s}$ | = | number of lots (dozens) of Space Rays to produce next week |
| $x_{p}$ | = | number of lots (dozens) of Phasers to produce next week |

| | | | | | | |
| --- | --- | --- | --- | --- | --- | --- |
| $\max$ | $8x_{s}$ | $+$ | $5x_{p}$ | | | |
| s.t. | $2x_{s}$ | $+$ | $1x_{p}$ | $\le$ | $1200$ | {plastic pounds} |
| | $3x_{s}$ | $+$ | $4x_{p}$ | $\le$ | $2400$ | {minutes of production} |
| | $1x_{s}$ | $+$ | $1x_{p}$ | $\le$ | $800$ | {overall production limit} |
| | $1x_{s}$ | $-$ | $1x_{p}$ | $\le$ | $450$ | {mix of products produced} |
| | $x_{s}$ | | | $\ge$ | $0$ | {non-negativity} |
| | | | $x_{p}$ | $\ge$ | $0$ | {non-negativity} |

In [110]:
# Create the model object
m = gp.Model('galaxy_toys')

In [111]:
# Specify how to optimize and time limit (seconds)
m.ModelSense = GRB.MAXIMIZE
m.setParam('TimeLimit', 600)

# update the model
m.update()

Set parameter TimeLimit to value 600


In [112]:
# Create decision variables
# We tell the solver that the variables are continuous,
#   their names, and their lower bounds
x_s = m.addVar(vtype=GRB.CONTINUOUS, name='x_s', lb=0.0)
x_p = m.addVar(vtype=GRB.CONTINUOUS, name='x_p', lb=0.0)

# update the model
m.update()

In [113]:
# Add the objective function
m.setObjective(8 * x_s + 5 * x_p)
m.update()

In [114]:
# Add the constraints
# We can simply write out the constraints for the first parameter
# The second parameter names the constraint
m.addConstr(2*x_s + x_p <= 1200, name='plastic')
m.addConstr(3*x_s + 4*x_p <= 2400, name='labor')
m.addConstr(x_s + x_p <= 800, name='total_production')
m.addConstr(x_s - x_p <= 450, name='product_mix')
m.update()

In [115]:
# This method is undocumented and deprecated
# BUT I think it still works
m.display()

Maximize
  8.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\941354986.py:3: DeprecationWarning: Model.display() is deprecated
  m.display()


In [116]:
# solve
m.optimize()

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Model fingerprint: 0xbb372ecc
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 8e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
Presolve time: 0.00s
Presolved: 4 rows, 2 columns, 8 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.3000000e+31   5.250000e+30   1.300000e+01      0s
       4    5.0400000e+03   0.000000e+00   0.000000e+00      0s

Solved in 4 iterations and 0.00 seconds (0.00 work units)
Optimal objective  5.040000000e+03


### Getting the Results

We now want to get the results and print them in a little nicer format. 

In [117]:
# Get the results out
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

To generate the optimal profit of $5040.00, you should produce the following amounts:
   x_s = 480.0
   x_p = 240.0


### Sensitivity Analysis

We can easily get the reduced cost and the range of optimality for each variable. Similarly, we can extract the shadow price and the range feasibility for each constraint.

In [118]:
# Get the reduced cost and range of optimality for each variable
for v in m.getVars():
    print(f'{v.VarName} has a reduced cost of {v.RC}')
    print(f'   and a range of optimality from {v.SAObjLow} to {v.SAObjUp}')

x_s has a reduced cost of 0.0
   and a range of optimality from 3.75 to 10.0
x_p has a reduced cost of 0.0
   and a range of optimality from 4.0 to 10.666666666666668


In [119]:
# Get the shadow price and the range of feasibility for each constraint
for c in m.getConstrs():
    print(f'{c.constrName} has a shadow price of {c.pi}')
    print(f'   and a range of feasibility from {c.SARHSLow} to {c.SARHSUp}')
    print(f'   and a slack of {c.Slack}')
    print(f'   and a RHS of {c.RHS}')
    print(f'   and a final value of {m.getRow(c).getValue()}')

plastic has a shadow price of 3.4
   and a range of feasibility from 600.0 to 1350.0
   and a slack of 0.0
   and a RHS of 1200.0
   and a final value of 1200.0
labor has a shadow price of 0.4
   and a range of feasibility from 2050.0 to 2800.0
   and a slack of 0.0
   and a RHS of 2400.0
   and a final value of 2400.0
total_production has a shadow price of 0.0
   and a range of feasibility from 720.0 to inf
   and a slack of 80.0
   and a RHS of 800.0
   and a final value of 720.0
product_mix has a shadow price of 0.0
   and a range of feasibility from 240.0 to inf
   and a slack of 210.0
   and a RHS of 450.0
   and a final value of 240.0


### Modularized Functionality

Up at the top of this notebook, I created two functions to make it easier to extract the sensitivity analysis for both the variables, `sa_vars()`, and the constraints, `sa_constrs()`. Let's use those to get the sensitivity information out. First, we'll examine the docstring on the function and then use it.

In [120]:
help(sa_vars)

Help on function sa_vars in module __main__:

sa_vars(the_vars)
    This function takes the variables from a solved model
    and returns a pandas DataFrame where the index is the
    variable name and the columns are the (1) the resulting
    value of the variable, (2) the reduced cost, (3) the
    objective function coefficient, (4) the low end of the
    range of optimality, and (5) the high end of the range
    of optimality

    Parameters
    the_vars: the variables of the solved model (e.g., the
              result of m.getVars())

    Returns
    A pandas DataFrame object



In [121]:
# returns a DataFrame
# This will simply print it out
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s        480.0           0.0       8.0           3.75       10.000000
x_p        240.0           0.0       5.0           4.00       10.666667

### Range of Optimality

Suppose the objective function coefficient for space rays was really \$7. How does this change the solution? Can you answer that question using the sensitivity report? Do you have re-solve the problem witht the new data?

### Changing the Objective Function Coefficient for Space Rays

Let's go ahead and change the objective function coefficient from the original \\$8 to \\$7 and re-run the solver to see what we get.

In [122]:
# Change the objective function
m.setObjective(7 * x_s + 5 * x_p)
m.update()
m.display()

Maximize
  7.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\1929794627.py:4: DeprecationWarning: Model.display() is deprecated
  m.display()


In [123]:
# Solve the updated model with $7
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 7e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    4.5600000e+03   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  4.560000000e+03
To generate the optimal profit of $4560.00, you should produce the following amounts:
   x_s = 480.0
   x_p = 240.0


In [124]:
# Look at sensitivity analysis for variables again
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s        480.0           0.0       7.0           3.75       10.000000
x_p        240.0           0.0       5.0           3.50        9.333333

### Reduced Costs

We want to change the objective function coefficient of Space Rays to \\$2 and re-solve the model. Note that \\$2 is outside the range of optimality, so we need to re-solve the model to find the optimal solution.

In [125]:
# Change the objective function
m.setObjective(2 * x_s + 5 * x_p)
m.update()
m.display()

Maximize
  2.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\2300582346.py:4: DeprecationWarning: Model.display() is deprecated
  m.display()


In [126]:
# Solve the updated model with $2
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [2e+00, 5e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.8000000e+30   1.600000e+30   2.800000e+00      0s
       1    3.0000000e+03   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.00 seconds (0.00 work units)
Optimal objective  3.000000000e+03
To generate the optimal profit of $3000.00, you should produce the following amounts:
   x_s = 0.0
   x_p = 600.0


In [127]:
# Look at sensitivity analysis for variables again
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s          0.0         -1.75       2.0           -inf            3.75
x_p        600.0          0.00       5.0       2.666667             inf

We see that the reduced cost of $x_{s}$ is $-1.75$. This implies that the profit coefficient for Space Rays would have to increase by \\$1.75, which would be \\$2 + \\$1.75 = \\$3.75, before it becomes economically feasible to produce Space Rays. (Thought exercise: does that number look familiar?)

Another way to think of the reduced cost is: 'What if I were forced to produce one unit (one lot or one dozen) of Space Rays? What would happen to my profit?' Let's add another constraint to the model that forces one unit of Space Rays production and see the results. 

In [128]:
# Add a new constraint that says: x_s >= 1
# You could do this two ways:
# 1. as a structural constraint -- doing this one
# 2. by changing the lower bound on x_s
m.addConstr(x_s >= 1, name='force_space_ray_production')
m.update()
m.display()

Maximize
  2.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450
  force_space_ray_production: x_s >= 1


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\2809422579.py:7: DeprecationWarning: Model.display() is deprecated
  m.display()


In [129]:
# Solve and display results
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 5 rows, 2 columns and 9 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [2e+00, 5e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.0000000e+03   1.000000e+00   0.000000e+00      0s
       1    2.9982500e+03   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.00 seconds (0.00 work units)
Optimal objective  2.998250000e+03
To generate the optimal profit of $2998.25, you should produce the following amounts:
   x_s = 1.0
   x_p = 599.25


### Range of Feasibility

Suppose we had one additional pound of plastic that could be used for our production cycle. What does this additional resource do to our solution? Can you answer that with the sensitivity report?

First, let's go back to the original model. We need to reset the objective function coefficient for space rays and remove the constraint that forced the production of space rays. Then, let's change the right-hand-side of the plastic constraint and re-solve the problem.

In [130]:
# See what the current model looks like
m.display()

Maximize
  2.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450
  force_space_ray_production: x_s >= 1


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\818556226.py:2: DeprecationWarning: Model.display() is deprecated
  m.display()


#### A Quick Aside

If (and when) the method `.display()` goes away, you could do the following. Write the model to a temporary file in `.lp` format, then read it back in and print it. This approach is a **bad** idea for larger models. See below for an example. 

In [131]:
# One work around is to write the .lp file 
# to disk and and read it back in and 
# print it out ... not recommended for large models
m.write('./tmp.lp')
with open('./tmp.lp') as f:
    print(f.read())

\ Model galaxy_toys
\ LP format - for model browsing. Use MPS format to capture full model detail.
Maximize
  2 x_s + 5 x_p
Subject To
 plastic: 2 x_s + x_p <= 1200
 labor: 3 x_s + 4 x_p <= 2400
 total_production: x_s + x_p <= 800
 product_mix: x_s - x_p <= 450
 force_space_ray_production: x_s >= 1
Bounds
End



In [132]:
# Let's change objective function back to the original
m.setObjective(8 * x_s + 5 * x_p)
# And delete the force space rays production constraint
# Get the constraint by name and remove it
c = m.getConstrByName('force_space_ray_production')
if c:  # Check if the constraint exists
    m.remove(c)

m.update()
m.display()

Maximize
  8.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\3990297479.py:10: DeprecationWarning: Model.display() is deprecated
  m.display()


In [133]:
# Solve the model
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Model fingerprint: 0xbb372ecc
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 8e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
Presolve time: 0.00s
Presolved: 4 rows, 2 columns, 8 nonzeros

       0    1.3000000e+31   5.250000e+30   1.300000e+01      0s
       4    5.0400000e+03   0.000000e+00   0.000000e+00      0s

Solved in 4 iterations and 0.00 seconds (0.00 work units)
Optimal objective  5.040000000e+03
To generate the optimal profit of $5040.00, you should produce the following amounts:
   x_s = 480.0
   x_p = 240.0


### Getting Sensitivity Information for Constraints

You can use the function defined at the top of this notebook, `sa_constrs()`, to retrieve the sensitivity analysis for the constraints. The returned value from the function is a pandas `DataFrame`. 

In [134]:
# See the docstring for the function
help(sa_constrs)

Help on function sa_constrs in module __main__:

sa_constrs(the_constrs)
    This function takes the constraints from a solved model
    and returns a pandas DataFrame where the index is the
    constraint name and the columns are the (1) indication
    if the constraint is binding or non-binding, (2) the
    value of the constraint, (3) the RHS of the constraint,
    (4) the slack, (5) the shadow price, (6) the low end of
    the range of feasibility, and (5) the high end of the
    range of feasibility

    Parameters
    the_constrs: the constraints of the solved model (e.g., the
              result of m.getConstrs())

    Returns
    A pandas DataFrame object



In [135]:
sa_constrs(m.getConstrs())

binding?  final_value     RHS  slack  shadow_price  \
plastic               binding       1200.0  1200.0    0.0           3.4   
labor                 binding       2400.0  2400.0    0.0           0.4   
total_production  non-binding        720.0   800.0   80.0           0.0   
product_mix       non-binding        240.0   450.0  210.0           0.0   

                  range_feas_low  range_feas_high  
plastic                    600.0           1350.0  
labor                     2050.0           2800.0  
total_production           720.0              inf  
product_mix                240.0              inf

In [136]:
# We can pull out a specific constraint by its name
# Let's get plastic and store it a Python variable named plastic
plastic = m.getConstrByName('plastic')
print(plastic.RHS)

1200.0


In [137]:
# Now update the RHS with one additional pound of plastic
plastic.RHS = 1201

# Update the model and solve it
m.update()
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 8e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    5.0434000e+03   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  5.043400000e+03
To generate the optimal profit of $5043.40, you should produce the following amounts:
   x_s = 480.8
   x_p = 239.39999999999998


In [138]:
# sensitivity analysis for variables
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s        480.8           0.0       8.0           3.75       10.000000
x_p        239.4           0.0       5.0           4.00       10.666667

In [139]:
# sensitivity analysis for constraints
sa_constrs(m.getConstrs())

binding?  final_value     RHS  slack  shadow_price  \
plastic               binding       1201.0  1201.0    0.0           3.4   
labor                 binding       2400.0  2400.0    0.0           0.4   
total_production  non-binding        720.2   800.0   79.8           0.0   
product_mix       non-binding        241.4   450.0  208.6           0.0   

                  range_feas_low  range_feas_high  
plastic               600.000000           1350.0  
labor                2052.333333           2799.0  
total_production      720.200000              inf  
product_mix           241.400000              inf

### Multiple Optimal Solutions

Suppose the model has multiple optimal solutions. Can we find them? Let's change the objective function coefficient of space rays to \\$3.75 and re-solve the model. (We also need to change the plastic constraint's RHS back to 1200.) 

In [140]:
# See the current model
m.display()

Maximize
  8.0 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1201
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\143021986.py:2: DeprecationWarning: Model.display() is deprecated
  m.display()


In [141]:
# reset the rhs of the plastic constraint to 1200
print(f'Current RHS of plastic constraint is {plastic.rhs}')

plastic.rhs = 1200

# update the model
m.update()

print(f'After changing RHS of back to original is {plastic.rhs}')

Current RHS of plastic constraint is 1201.0
After changing RHS of back to original is 1200.0


In [142]:
# Change the objective function
m.setObjective(3.75 * x_s + 5 * x_p)
m.update()
m.display()

C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\463068037.py:4: DeprecationWarning: Model.display() is deprecated
  m.display()


Maximize
  3.75 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


In [143]:
# Solve
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [4e+00, 5e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.0000000e+03   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  3.000000000e+03
To generate the optimal profit of $3000.00, you should produce the following amounts:
   x_s = 480.0
   x_p = 240.0


In [144]:
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s        480.0           0.0      3.75          3.750            10.0
x_p        240.0           0.0      5.00          1.875             5.0

#### Multiple Optimal Solutions?

I am going to claim that this current model has multiple optimal solutions. One way to determine if that is true is to plug all of the corner points (we only had 5) into the objective function, $3.75x_{s} + 5x_{p}$, and see if any of the other corner points give us \\$3,000. (You can go back to where we visualized the model in a previous notebook and found all the corner points.)

1. (0, 0) = \\$0
2. (450, 0) = \\$1,687.50
3. (550, 100) = \\$2,562.50
4. (480, 240) = \\$3,000
5. (0, 600) = \\$3,000

So, yes, there are multiple optimal solutions to this model. Another clue that an alternate optimal solution exists can be seen in the sensitivity information: the objective function coefficient is the same as the low end of the range of optimality for $x_{s}$.

Sometimes, in the sensitivity report we will have the final value of a variable is 0 and its reduced cost is also 0. When this occurs, it is another clue that we may have multiple optimal solutions. 

Can we force the solver to find the other corner point that is optimal? Yes. We can change the objective function into a constraint and then maximize $x_{p}$. (We are doing this only because we know that the alternate solution has a higher value for $x_{p}$.) Let's try it.

In [145]:
# See the current model
m.display()

Maximize
  3.75 x_s + 5.0 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\143021986.py:2: DeprecationWarning: Model.display() is deprecated
  m.display()


In [146]:
# Turn the objective function into a constraint
m.addConstr(3.75*x_s + 5*x_p == 3000, name='set_obj_value_output')

# Change the objectiv function
m.setObjective(x_p)

m.update()
m.display()

Maximize
  x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450
  set_obj_value_output: 3.75 x_s + 5.0 x_p = 3000


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\796021131.py:8: DeprecationWarning: Model.display() is deprecated
  m.display()


In [147]:
# solve
m.optimize()

# get the results
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 5 rows, 2 columns and 10 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 5e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 3e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.2000000e+30   1.600000e+30   1.200000e+00      0s
       1    6.0000000e+02   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.00 seconds (0.00 work units)
Optimal objective  6.000000000e+02
   x_s = 0.0
   x_p = 600.0


In [148]:
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s          0.0         -0.75       0.0           -inf            0.75
x_p        600.0          0.00       1.0           -0.0             inf

### Multiple Optimal Solutions

Let's consider the model where space rays bring in \\$9 of profit and phasers bring in \\$4.50 of profit. I will claim that this model also has multiple optimal solutions. Let's adjust our model and see if we can find multiple optimal solutions to it.

In [149]:
# See current model
m.display()

Maximize
  x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\1768218809.py:2: DeprecationWarning: Model.display() is deprecated
  m.display()


  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450
  set_obj_value_output: 3.75 x_s + 5.0 x_p = 3000


In [150]:
# Need to delete the 'set_obj_value_output' constraint
c = m.getConstrByName('set_obj_value_output')
if c:  # Check if the constraint exists
    m.remove(c)
    
# and change the objective function
m.setObjective(9 * x_s + 4.5 * x_p)
m.update()
m.display()

Maximize
  9.0 x_s + 4.5 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\1854959763.py:9: DeprecationWarning: Model.display() is deprecated
  m.display()


In [151]:
# solve
m.optimize()

# get the results
print(f'To generate the optimal profit of ${m.ObjVal:0.2f}, you should produce the following amounts:')
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 9e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    5.6250000e+30   3.375000e+30   5.625000e+00      0s
       2    5.4000000e+03   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.00 seconds (0.00 work units)
Optimal objective  5.400000000e+03
To generate the optimal profit of $5400.00, you should produce the following amounts:
   x_s = 550.0
   x_p = 100.0


In [152]:
# Get the reduced cost and range of optimality for each variable
sa_vars(m.getVars())

final_value  reduced_cost  obj_coef  range_opt_low  range_opt_high
x_s        550.0           0.0       9.0            9.0             inf
x_p        100.0           0.0       4.5           -9.0             4.5

In [153]:
# get shadow prices and range of feasibility
sa_constrs(m.getConstrs())

binding?  final_value     RHS  slack  shadow_price  \
plastic               binding       1200.0  1200.0    0.0           4.5   
labor             non-binding       2050.0  2400.0  350.0           0.0   
total_production  non-binding        650.0   800.0  150.0           0.0   
product_mix           binding        450.0   450.0    0.0           0.0   

                  range_feas_low  range_feas_high  
plastic                    900.0           1350.0  
labor                     2050.0              inf  
total_production           650.0              inf  
product_mix                240.0            600.0

#### How to Find Other Optimal Solutions

That answer only showed a single solution. We can ask Gurobi to find multiple solutions and then look at those.

In [154]:
# do a systematic search for the k-best solutions
m.setParam(GRB.Param.PoolSearchMode, 2)

# Update the model
m.update()

# Solve
m.optimize()

# Get the status
status = m.Status
if status in (GRB.INF_OR_UNBD, GRB.INFEASIBLE, GRB.UNBOUNDED):
    print('The model cannot be solved because it is infeasible or unbounded')

if status != GRB.OPTIMAL:
    print('Optimization was stopped with status ' + str(status))

if status == GRB.OPTIMAL:
    print(f'Optimization found an OPTIMAL answer with the status of {status}')

# Print number of solutions stored
nSolutions = m.SolCount
print('Number of solutions found: ' + str(nSolutions))

Set parameter PoolSearchMode to value 2
Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600
PoolSearchMode  2

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 9e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  5.400000000e+03
Optimization found an OPTIMAL answer with the status of 2
Number of solutions found: 1


#### Why Only 1 Solution?

When the variables are continuous and you have multiple optimal solutions, you will in fact have an **infinite** number of optimal solutions. Therefore, the solver will not try to search for more than one. Which one did it find?

In [155]:
for v in m.getVars():
    print(f'   {v.VarName} = {v.X}')

   x_s = 550.0
   x_p = 100.0


#### Another Possible Approach?

Because the optimal solution found above is integer, you could try converting both variables to integer instead of continuous and re-solving the updated model. When dealing with integer or binary variables, Gurobi will try to find other optimal solutions.

In [156]:
# change the variable types to integer
for var in m.getVars():
    print(f'Variable {var.VarName} is currently of type {var.VTYPE}')
    var.setAttr(GRB.Attr.VType, GRB.INTEGER)

# update the model for changes to take effect
m.update()
print('After setting the variables to integer ...')
for var in m.getVars():
    print(f'Variable {var.VarName} is currently of type {var.VTYPE}')

Variable x_s is currently of type C
Variable x_p is currently of type C
After setting the variables to integer ...
Variable x_s is currently of type I
Variable x_p is currently of type I


In [157]:
# see current model
m.display()

Maximize
  9.0 x_s + 4.5 x_p
Subject To
  plastic: 2.0 x_s + x_p <= 1200
  labor: 3.0 x_s + 4.0 x_p <= 2400
  total_production: x_s + x_p <= 800
  product_mix: x_s + -1.0 x_p <= 450
General Integers
  ['x_s', 'x_p']


C:\Users\zungz\AppData\Local\Temp\ipykernel_13004\2220004513.py:2: DeprecationWarning: Model.display() is deprecated
  m.display()


In [158]:
# Solve
m.optimize()

# Print number of solutions stored
nSolutions = m.SolCount
print(f'Number of solutions found: {nSolutions}')

Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (win64 - Windows 11.0 (26100.2))

CPU model: AMD Ryzen 9 8945HS w/ Radeon 780M Graphics, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Non-default parameters:
TimeLimit  600
PoolSearchMode  2

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Model fingerprint: 0x55205400
Variable types: 0 continuous, 2 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [5e+00, 9e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+02, 2e+03]
Found heuristic solution: objective 5229.0000000
Presolve removed 1 rows and 0 columns
Presolve time: 0.00s
Presolved: 3 rows, 2 columns, 6 nonzeros
Variable types: 0 continuous, 2 integer (0 binary)

Root relaxation: objective 5.400000e+03, 1 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth In

In [159]:
# Print objective value and values of the decision varaibles of solutions
for soln_num in range(m.getAttr(GRB.Attr.SolCount)):
    m.Params.SolutionNumber = soln_num
    print(f'solution {soln_num} has obj fn value of {m.PoolObjVal}')
    for v in m.getVars():
        # Need to use v.Xn to the value of the variable v for this solution
        print(f'   {v.VarName} = {v.Xn}')
    

solution 0 has obj fn value of 5400.0
   x_s = 550.0
   x_p = 100.0
solution 1 has obj fn value of 5400.0
   x_s = 482.0
   x_p = 236.00000000000003
solution 2 has obj fn value of 5400.0
   x_s = 516.0
   x_p = 168.0
solution 3 has obj fn value of 5400.0
   x_s = 508.0
   x_p = 184.0
solution 4 has obj fn value of 5400.0
   x_s = 499.0
   x_p = 202.0
solution 5 has obj fn value of 5400.0
   x_s = 491.0
   x_p = 218.0
solution 6 has obj fn value of 5400.0
   x_s = 514.0
   x_p = 172.0
solution 7 has obj fn value of 5400.0
   x_s = 504.0
   x_p = 192.0
solution 8 has obj fn value of 5400.0
   x_s = 487.0
   x_p = 226.0
solution 9 has obj fn value of 5400.0
   x_s = 489.0
   x_p = 222.0


### Success!

When using integer variables, we can get multiple optimal solutions with Gurobi. Note that within a solution you need to use **`v.Xn`**, where `v` is the variable, to get the values of the decision variables. 

**&copy; 2024 - Present: Matthew D. Dean, Ph.D.   
Clinical Professor of Business Analytics at William \& Mary.**